strawberry/contrib/django/tests/test_model_type.py

```diff
@@ -0,0 +1,195 @@
+            return TestModel(name="modeltype")
+
+    schema = strawberry.Schema(query=Query)
+
+    response = await execute(
+        query="""
+        query {
+            modelType {
+                name
+                extraInfo
+            }
+        }
+    """,
+        schema=schema,
+    )
+
+    assert response.data["modelType"]["name"] == "modeltype"
+    assert response.data["modelType"]["extraInfo"] == "more info"
+
+
+@pytest.mark.asyncio
+async def test_model_type_can_override_type_fields():
+    @model_type(model=TestModel, fields=["name"])
+    class TestModelType:
+        @strawberry.field
+        def name(self, info) -> str:
+            return "joe"
+
+    @strawberry.type
+    class Query:
+        @strawberry.field
+        def model_type(root, info) -> TestModelType:
+            return TestModel(name="modeltype")
+
+    schema = strawberry.Schema(query=Query)
+
+    response = await execute(
+        query="""
+        query {
+            modelType {
+                name
+            }
+        }
+    """,
+        schema=schema,
+    )
+
+    assert response.data["modelType"]["name"] == "joe"
+
+
+@pytest.mark.asyncio
+async def test_model_type_only_fields():
+    @model_type(model=TestModel, fields=["name"])
+    class TestModelType:
+        pass
+
+    assert "name" in TestModelType.field.fields
+    assert "id" not in TestModelType.field.fields
+    assert "secret" not in TestModelType.field.fields
+
+
+@pytest.mark.asyncio
+@pytest.mark.django_db
+async def test_model_type_list():
+    @model_type(model=TestModel, fields="__all__")
+    class TestModelType:
+        @strawberry.field
+        def extra_info(self, info) -> str:
+            return "more info"
+
+    @strawberry.type
+    class Query:
+        @strawberry.field
+        def model_type_list(root, info) -> List[TestModelType]:
+            return TestModel.objects.all()
+
+    schema = strawberry.Schema(query=Query)
+
+    TestModel.objects.create(name="modeltype")
+    TestModel.objects.create(name="modeltype")
+    TestModel.objects.create(name="modeltype")
+
+    response = await execute(
+        query="""
+        query {
+            modelTypeList {
+                id
+                name
+                extraInfo
+            }
+        }
+    """,
+        schema=schema,
+    )
+
+    assert len(response.data["modelTypeList"]) == 3
+
+
+@pytest.mark.asyncio
+@pytest.mark.django_db
+async def test_model_input_type():
+    @model_type(model=TestModel, fields="__all__")
+    class TestModelType:
+        extra: str
+
+    @model_type(model=TestModel, is_input=True, fields="__all__")
+    class TestModelInputType:
+        pass
+
+    @strawberry.type
+    class Query:
+        hello: str = "world"
+
+    @strawberry.type
+    class Mutation:
+        @strawberry.field
+        def create_test_model(self, info, test: TestModelInputType) -> TestModelType:
+            name = getattr(test, "name")
```

Comment:
mmh, why getattr here? :)

Reply:
`mypy` complains that `"TestModelInputType" has no attribute "name"` if it's accessed directly. Do I need to make some adjustments to our `mypy` plugin for this to work?

Reply:
will check this tomorrow, probably!

Reply:
@patrick91 After looking into this, it seems like `mypy` support might be difficult to add here. The `model_type` decorator adds attributes to the class dynamically using `setattr`, but this isn't compatible with `mypy`'s static analysis. A [similar issue](https://github.com/python/mypy/issues/5719#issuecomment-426427083) in the mypy repo notes that:

> ... mypy doesn't run your code, it only reads it, and it only looks at the types of attributes, not at actions like setattr().

We could potentially write a custom plugin for this (as the poster of that issue did), but that seems overly complex in this situation.